### PR TITLE
[plugin] Tidy up battery statistics

### PR DESCRIPTION
**Pull request by zwim**
Tidy the entries and add method to change displayed lines in the battery-statistics plugin

Before (huge font, %-symbol scattered all around the screen, decimal hours :( ):
grafik

After (Each number with unit, consistent formatting, duration in the form 2d12h23m, changeable number of lines):
grafik

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8491)

### CODE REVIEW

**Comment by zwim** [comment]
No problem to drop `batterystat_items_per_page`, we can just use 23 and all fits on one page. 
Maybe then we could also drop the bottom navigation line (as we only have one page) and make the text a bit bigger.

File: frontend/ui/widget/keyvaluepage.lua [hunk 1]
```diff
@@ -337,6 +337,7 @@ local KeyValuePage = InputContainer:new{
     -- aligment of value when key or value overflows its reserved width (for
     -- now: 50%): "left" (stick to key), "right" (stick to scren right border)
     value_overflow_align = "left",
+    -- single_page = nil,
```
Comment (poire-z) [inline frontend/ui/widget/keyvaluepage.lua:340]: Leave it uncommented (like we have above for width/height. Having it commented out, we may think it's an old option that is no longer valid.

File: frontend/ui/widget/keyvaluepage.lua [hunk 2]
```diff
@@ -519,18 +526,34 @@ function KeyValuePage:init()
...
+                align = "left",
+                self.title_bar,
+                VerticalSpan:new{ width = span_height },
+                self.main_content,
+            },
+            page_return,
+            footer,
+        }
+    else
+        content = OverlapGroup:new{
```
Comment (poire-z) [inline frontend/ui/widget/keyvaluepage.lua:544]: No need for this if/else and big duplication - if I get it right that only footer is there or not.
As footer is last, no issue it being nil :)

And even if it wasn't last, you could have use:
```lua
previous_widget,
footer or VerticalSpan:new{width=0},
next_widget,
```

File: frontend/ui/widget/keyvaluepage.lua [hunk 3]
```diff
@@ -467,20 +381,112 @@ function KeyValuePage:init()
...
-    local page_return = BottomContainer:new{
-        dimen = self.inner_dimen:copy(),
-        WidgetContainer:new{
-            dimen = Geom:new{
-                w = self.inner_dimen.w,
-                h = self.return_button:getSize().h,
+    local footer -- will stay nil if self.single_page is true
+    if self.single_page then
+        -- group for page info
+        local chevron_left = "chevron.left"
```
Comment (poire-z) [inline frontend/ui/widget/keyvaluepage.lua:387]: I guess you wanted to write `if NOT self.single_page` :)

(That's a lot of changes/diff because (only?) of some indentation changes. I always have some moral dilemna when that happens to me :) I would  probably have gone in this case with just creating all these whether single_page or not, and just not use it if single_page (with may be a footer:free() to be sure) - as single_page is a super rare use case for now.)

**Comment by Frenzie** [comment]
The single page thing you mention is to shrink the font size to ensure it'll fit or how am I supposed to interpret that?

**Comment by zwim** [comment]
@Frenzie : The single_page thing is for omitting the footer navigation (<<  <  1 of 1 > >>). First because it is useless on one page and second we have more space for out 21 lines of text (tex is bigger).

**Comment by poire-z** [comment]
Oh, right, that's what it does :) So, better name it really as what it does: `show_footer = true`.

(We could also checked if all fits on one page, but that's a tad complicated - and may be surprising in some KVPage usage where the nb of items may vary.)

**Comment by zwim** [comment]
`show_footer = false` does not imply that there is just one page. What would two pages without navigation help?

I would stay with `single_page` (as it is clear that we then definitively don't need a navigation).
If one wants navigation (for what reasons ever), he  can use the widget as usual.

**Comment by zwim** [comment]
>(We could also checked if all fits on one page, but that's a tad complicated - and may be surprising in some KVPage usage where the nb of items may vary.)

The bad thing would be, that the height of the items would change then. But we could think about (in a different PR), if we just don't draw the then useless navigation footer at all or grey him out, or just say 1 of 1 without the chevrons).

**Comment by poire-z** [comment]
> `show_footer = false` does not imply that there is just one page. What would two pages without navigation help?
> 
> I would stay with `single_page` (as it is clear that we then definitively don't need a navigation. Nobody expects navigation then). If one wants navigation (for what reasons ever), he can use the widget as usual.

I get that you need X (no footer) when Y (showing all items on a single page) - and that X without Y makes little sense (we could still browse pages with swipes, but ok :). But that does not really justify calling X (no footer) Y (single page) :)
Moreover, you don't prevent with single_page=true from actually providing 25 items and giving  items_per_page=15 : there, I'll get a single page with 15 items, without navigation, an no ability of seeing the 2nd page, or even knowing about it.
So, either show_footer (does what it says).
Or have single_page does a bit more: count the real nb of items, use that as items_per_page (so, no need for caller to provide it), and put all of them really on a single page with no footer (does what it says too :)

> The bad thing would be, that the height of the items would change then. But we could think 

Yes, it would be bad, so no need to give it any thought :)

**Comment by zwim** [comment]
Convinced :)

**Comment by poire-z** [comment]
> Convinced :)

Damned, I was hoping I would have you convinced... with the other alternative (single_page doing more stuff :)
(So other use casse could just provide the single_page=true option, and would not need to count their items - I would find this quite elegant :)

**Comment by zwim** [comment]
> > Convinced :)
> 
> Damned, I was hoping I would have you convinced... with the other alternative (single_page doing more stuff :) (So other use casse could just provide the single_page=true option, and would not need to count their items - I would find this quite elegant :)

Yes, on the long term we might - no we will - end up with the single_page (and a show_footer?) solution. When I look to the system statistics ...
Always counting the items, maybe loosing the last one on the next (invisible page ).

As this PR is ready from my side (and it advances the plugin ), I  would want to merge it as is.
As soon as someone/I have done the single_page thing, it is easy an going change: single_page instead show_footer and remove items_per_page.

**Comment by poire-z** [comment]
I'm fine with merging as it is. We'll just be vigilant that no one is using your show_footer (or items_per_page, although this might have some use) that may be removed.
(Or you could remove them now, and just leave for some time with battery statistics with a footer - as you wish.)

> with the other alternative (single_page doing more stuff :)
(So other use casse could just provide the single_page=true option, and would not need to count their items - I would find this quite elegant :)

Might need some thinking though.
I remember I set my KVPage values_per_page so one panel (probably Current statistics) would fit all in a single page (but I see I have room for one more row, so may be it was something else - or it's the separators that adds a bit of height....)
My idea just above is that we might want to throw a single_page to these statistics KVPage, or to the Book information KVPage - but then, if there are less items, they would get a bigger font, and may be the key/values would be too truncated.
So, you might want to use items_per_page = math.min(computed nb of items provided, user preferred items_per_page), so we would still get them on a single page - but not filling that page?
But on the flip side, a user keyvalues_per_page might mean a font size that he can read, and making it smaller to fit more items on a page might be bothering...

**Comment by zwim** [comment]
An unrelated thought

>>The bad thing would be, that the height of the items would change then. But we could think

> Yes, it would be bad, so no need to give it any thought :)

What would be bad if we make the chevrons very light if there is only one page. Or in the case we have more pages and we are on the last/first page lighten the corresponding ones?
I think it would be intuitive, I see that I dont have to tap on the light ones. In the current state  I have to read 1 of 5 and must think: Oh I am on the first page, dont touch me.

**Comment by poire-z** [comment]
> What would be bad if we make the chevrons very light if there is only one page. Or in the case we have more pages and we are on the last/first page lighten the corresponding ones?
I think it would be intuitive, I see that I dont have to tap on the light ones. 

Mhhh, but isn't it already the case ?! I see my chevrons gray in Book information. I see them black (but not the left ones in the first page) of Dictionary history.
May be this gray is too dark for you or your device's eInk ?

**Comment by NiLuJe** [comment]
We already had a few discussions about what to do with the nav bar/buttons on single-page screens and/or last screens back when we re-designed all widgets with those to look the same.

TL;DR: Current state was the result of a consensus ;).

File: plugins/batterystat.koplugin/main.lua [hunk 4]
```diff
@@ -4,12 +4,15 @@ local Dispatcher = require("dispatcher")
 local KeyValuePage = require("ui/widget/keyvaluepage")
 local LuaSettings = require("luasettings")
 local PowerD = require("device"):getPowerDevice()
+local SpinWidget = require("ui/widget/spinwidget")
```
Comment (hius07) [inline plugins/batterystat.koplugin/main.lua:7]: Not used anymore

Comment (zwim) [reply plugins/batterystat.koplugin/main.lua]: Thanks.

**Comment by zwim** [comment]
The last commit drops `show_footer` and brings `single_page` and `min_nb_items`.

If `single_page` is nil, everything stays as it was.

If `single_page` is set and `min_nb_items` is not set  all items are squeezed on one page. It is up to us, that there are not too many items on a page (as they might get tiny).

If `single_page` is set and `min_nb_items` is set, then the items don't get too large (and so they don't get truncated).

File: frontend/ui/widget/keyvaluepage.lua [hunk 5]
```diff
@@ -495,11 +501,21 @@ function KeyValuePage:init()
...
                          - 2*Size.line.thick
                             -- account for possibly 2 separator lines added
 
-    self.items_per_page = G_reader_settings:readSetting("keyvalues_per_page") or self:getDefaultKeyValuesPerPage()
+    local force_items_per_page
+    if self.single_page then
+        if not self.min_nb_items then
+            force_items_per_page = #self.kv_pairs
+        else
+            force_items_per_page = math.max(#self.kv_pairs, self.min_nb_items)
```
Comment (poire-z) [inline frontend/ui/widget/keyvaluepage.lua:513]: I'm not sure `min_nb_items` is a good idea: when would we developpers feel like we have to use it, and with which hardcoded value ? and if not hardcoded, a value based on which other thingy ?

I would probably just use:
```lua
force_items_per_page = math.max(#self.kv_pairs,
    G_reader_settings:readSetting("keyvalues_per_page") or self:getDefaultKeyValuesPerPage() )
```
That is: we know the user is fine reading with its chosen `G_reader_settings:readSetting("keyvalues_per_page")`, so it's ok using it if we have less items - he will get some blank area at bottom of the page, no real need to zoom everything.
(But with `self.single_page` we just hope he will be fine if we go a little smaller to fit everything on one page.)

**Comment by zwim** [comment]
Yes, indeed. "keyvalues_per_page" is a real good default, much better than `min_nb_items`. So the decision is delegated to the user and the requirements the device has. :)
